python/ray/autoscaler/autoscaler.py

```diff
@@ -199,6 +202,62 @@ def _info(self):
         }
 
 
+class NodeLauncher(threading.Thread):
+    def __init__(self, queue, pending, config, *args, **kwargs):
+        self.queue = queue
+        self.pending = pending
+        self.provider = get_node_provider(config["provider"],
+                                          config["cluster_name"])
+        super().__init__(*args, **kwargs)
+
+    def _launch_node(self, config, count):
+        tag_filters = {TAG_RAY_NODE_TYPE: "worker"}
+        before = self.provider.nodes(tag_filters=tag_filters)
+        launch_hash = hash_launch_conf(config["worker_nodes"], config["auth"])
+        self.provider.create_node(
+            config["worker_nodes"], {
+                TAG_RAY_NODE_NAME: "ray-{}-worker".format(
+                    config["cluster_name"]),
+                TAG_RAY_NODE_TYPE: "worker",
+                TAG_RAY_NODE_STATUS: "uninitialized",
+                TAG_RAY_LAUNCH_CONFIG: launch_hash,
+            }, count)
+        after = self.provider.nodes(tag_filters=tag_filters)
+        if set(after).issubset(before):
+            print("Warning: No new nodes reported after node creation")
+
+    def run(self):
+        while True:
+            config, count = self.queue.get()
+            try:
+                self._launch_node(config, count)
+            finally:
+                self.queue.task_done()
```

Comment:
Do we use the task_done()?

Reply:
I added this in case we wanted to join() in the future but can get rid of it.